[HLSL] Add __builtin_hlsl_is_scalarized_layout_compatible

llvm-beanz opened the pull request:
HLSL tends to rely pretty aggressively on scalarization occuring in the complier, which allows for some relaxed language behaviors when types are fully sclarized to equivalent scalar representations.

This change adds a new queryable trait builtin for scalarized layout compatability.

Resolves #100614

llvm-beanz committed:
[HLSL] Add __is_scalarized_layout_compatible

HLSL tends to rely pretty aggressively on scalarization occuring in the
complier, which allows for some relaxed language behaviors when types
are fully sclarized to equivalent scalar representations.

This change adds a new queryable trait builtin for scalarized layout
compatability.

Resolves #100614

hekota reviewed (approved):
LGTM!

MitalAshok commented:
Is this the expected behaviour:

```c++
struct EmptyPadding : Empty { Empty e; int X; };
struct Repeated : Empty, DerivedEmpty { int X; }; // expected-warning {{inaccessible due to ambiguity}}
_Static_assert(sizeof(EmptyPadding) == 8, "");
_Static_assert(sizeof(Repeated) == 8, "");
_Static_assert(sizeof(NotEmpty) == 4, "");
_Static_assert(__is_scalarized_layout_compatible(EmptyPadding, NotEmpty), "");
_Static_assert(__is_scalarized_layout_compatible(Repeated, NotEmpty), "");
```

I.e., is the "scalarization" expected to get rid of padding? A similar issue happens with padding introduced by over-alignment.

Also see https://github.com/llvm/llvm-project/issues/98310#issuecomment-2248326726: This might need to be renamed to `__builtin_is_scalarized_layout_compatible` to be consistent with new type traits. Unsure if this applies to HLSL or Clang-only extensions (I'll ask for confirmation), but the logic is that standard libraries can use `__is_scalarized_layout_compatible` and the compiler provides `__builtin_is_scalarized_layout_compatible`.

damyanp commented:
> HLSL tends to rely pretty aggressively on scalarization occuring in the complier, which allows for some relaxed language behaviors when types are fully sclarized to equivalent scalar representations.

Some typos:

* occuring --> occurring
* complier --> compiler
* sclarized --> scalarized

damyanp commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -1142,3 +1142,85 @@ bool SemaHLSL::CheckBuiltinFunctionCall(unsigned BuiltinID, CallExpr *TheCall) {
+    }
+    if (const auto *RT = dyn_cast<RecordType>(T)) {
+      const RecordDecl *RD = RT->getDecl();
+      if (RD->isUnion()) {
```
Can vector types be put inside unions, or are unions guaranteed to be scalar? Does this deserve a comment explaining it in the same way that there's one for vectors?

llvm-beanz replied:
Unions are tricky... we don't actually support them in HLSL officially yet, but I should probably make this work with Unions since they are likely to be a [202y feature](https://github.com/microsoft/hlsl-specs/blob/main/proposals/0004-unions.md).

damyanp commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -1142,3 +1142,85 @@ bool SemaHLSL::CheckBuiltinFunctionCall(unsigned BuiltinID, CallExpr *TheCall) {
+      for (const auto *FD : RD->fields())
+        FieldTypes.push_back(FD->getType());
+      // Reverse the newly added sub-range.
+      std::reverse(FieldTypes.begin(), FieldTypes.end());
```
Is it obvious to someone with more domain knowledge why this has to be reversed?

llvm-beanz replied:
It's because of the worklist processing, nothing domain specific. For example, given:

```hlsl
struct T {
  int X;
  float Y;
};

struct V {
  T t;
  double D;
};
```

When I build the worklist for `V` it does:
* Start with V : Worklist: {V} Result {}
* Iterate V's members: Worklist: { T, double } Result {}
* _Reverse added members_ : Worklist: { double, T } Result {}
* Pop `T` from back() and iterate it's members: Worklist : { double, int, float } Result {}
* _Reverse added members_ : Worklist: { double, float, int } Result {}
* Pop and record it in the result : Worklist: { double, float} Result { int }
* Pop and record it in the result : Worklist: { double } Result { int, float }
* Pop and record it in the result : Worklist: { } Result { int, float, double }

damyanp commented on clang/test/SemaHLSL/Types/Traits/ScalarizedLayoutCompatible.hlsl:
```diff
@@ -0,0 +1,132 @@
+  Mikey,
+};
+
+enum Mammals : uint {
```
No Rat or Splinter??

damyanp commented on clang/test/SemaHLSL/Types/Traits/ScalarizedLayoutCompatible.hlsl:
```diff
@@ -0,0 +1,132 @@
+_Static_assert(__is_scalarized_layout_compatible(Empty, DerivedEmpty), "");
+
+_Static_assert(__is_scalarized_layout_compatible(NotEmpty, DerivedNotEmpty), "");
+_Static_assert(__is_scalarized_layout_compatible(NotEmpty, DerivedEmptyNotEmptyBase), "");
```
test case for unions?

llvm-beanz commented:
> Is this the expected behaviour:
> 
> ```c++
> struct EmptyPadding : Empty { Empty e; int X; };
> struct Repeated : Empty, DerivedEmpty { int X; }; // expected-warning {{inaccessible due to ambiguity}}
> _Static_assert(sizeof(EmptyPadding) == 8, "");
> _Static_assert(sizeof(Repeated) == 8, "");
> _Static_assert(sizeof(NotEmpty) == 4, "");
> _Static_assert(__is_scalarized_layout_compatible(EmptyPadding, NotEmpty), "");
> _Static_assert(__is_scalarized_layout_compatible(Repeated, NotEmpty), "");
> ```
> 
> I.e., is the "scalarization" expected to get rid of padding? A similar issue happens with padding introduced by over-alignment.

Yes, the scalarized layout ignores padding. Because HLSL generally assumes that structures become fully scalarized with elements stored in registers we have way laxer requirements around structure layout for "local" memory.

> Also see [#98310 (comment)](https://github.com/llvm/llvm-project/issues/98310#issuecomment-2248326726): This should be renamed to `__builtin_is_scalarized_layout_compatible` to be consistent with new type traits.

Will do!

AaronBallman commented on clang/include/clang/Basic/TokenKinds.def:
```diff
@@ -658,6 +658,9 @@ KEYWORD(out                         , KEYHLSL)
 #include "clang/Basic/HLSLIntangibleTypes.def"
 
+// HLSL Type traits.
+TYPE_TRAIT_2(__is_scalarized_layout_compatible, IsScalarizedLayoutCompatible, KEYHLSL)
```
Can you prefix this with `__builtin_`? We have bad experiences with accidental naming conflicts (not that I think this will run into it), and we're starting to add a prefix now to help avoid that.

llvm-beanz committed:
Update to PR feedback

* Prefix the trait builtin with `__builtin`
* Add test cases for unions

llvm-beanz committed:
Fix formatting

AaronBallman commented on clang/lib/Sema/SemaExprCXX.cpp:
```diff
@@ -6188,6 +6189,23 @@ static bool EvaluateBinaryTypeTrait(Sema &Self, TypeTrait BTT, const TypeSourceI
                Info) == TemplateDeductionResult::Success;
   }
+  case BTT_IsScalarizedLayoutCompatible: {
+    if (!LhsT->isVoidType() && !LhsT->isIncompleteArrayType())
+      if (Self.RequireCompleteType(Lhs->getTypeLoc().getBeginLoc(), LhsT,
+                                   diag::err_incomplete_type))
+        return true;
```
```suggestion
    if (!LhsT->isVoidType() && !LhsT->isIncompleteArrayType() &&
        Self.RequireCompleteType(Lhs->getTypeLoc().getBeginLoc(), LhsT,
                                   diag::err_incomplete_type))
      return true;
```

AaronBallman commented on clang/lib/Sema/SemaExprCXX.cpp:
```diff
@@ -6188,6 +6189,23 @@ static bool EvaluateBinaryTypeTrait(Sema &Self, TypeTrait BTT, const TypeSourceI
+      if (Self.RequireCompleteType(Lhs->getTypeLoc().getBeginLoc(), LhsT,
+                                   diag::err_incomplete_type))
+        return true;
+    if (!RhsT->isVoidType() && !RhsT->isIncompleteArrayType())
+      if (Self.RequireCompleteType(Rhs->getTypeLoc().getBeginLoc(), RhsT,
+                                   diag::err_incomplete_type))
+        return true;
```
```suggestion
    if (!RhsT->isVoidType() && !RhsT->isIncompleteArrayType() &&
          Self.RequireCompleteType(Rhs->getTypeLoc().getBeginLoc(), RhsT,
                                   diag::err_incomplete_type))
      return true;
```

AaronBallman commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -1142,3 +1142,85 @@ bool SemaHLSL::CheckBuiltinFunctionCall(unsigned BuiltinID, CallExpr *TheCall) {
+  if (T1.isNull() || T2.isNull())
+    return false;
+
+  T1 = T1.getCanonicalType().getUnqualifiedType();
+  T2 = T2.getCanonicalType().getUnqualifiedType();
```
Do you have to care about things like `_Atomic` (that won't get stripped by these calls)?

llvm-beanz replied:
HLSL doesn't have `_Atomic` or anything similar (yet), so I'm kinda guessing here. My gut is that we probably _do_ want to treat `_Atomic int` as different from `int` because we may have different addressing and storage requirements for atomic values than non-atomic values which may not be accounted for by our aggressive scalarization. So, my feeling is that it is safer for us to say `_Atomic int` is different from `int` for this case.

I _really_ want to remove a bunch (or all) of the language features that are built around this scalarization though. I think much of the need for this will go away once we support user-defined conversion sequences in a thought out way, but we'll need to get user code to adapt which may take time.

AaronBallman replied:
SGTM, thanks!

llvm-beanz committed:
Update clang/lib/Sema/SemaExprCXX.cpp

Co-authored-by: Aaron Ballman <aaron@aaronballman.com>

llvm-beanz committed:
Rename builtin to start with __builtin_hlsl
../clang/test/SemaHLSL/Types/Traits/ScalarizedLayoutCompatible.hlsl
../clang/test/SemaHLSL/Types/Traits/ScalarizedLayoutCompatibleErrors.hls
l

llvm-beanz committed:
clang-format

llvm-beanz committed:
Merge remote-tracking branch 'origin/main' into
cbieneman/flattened-types

# Conflicts:
#	clang/lib/Headers/hlsl/hlsl_basic_types.h

llvm-beanz committed:
Update clang/lib/Sema/SemaExprCXX.cpp

Co-authored-by: Aaron Ballman <aaron@aaronballman.com>

llvm-beanz committed:
clang-format

AaronBallman reviewed (approved):
LGTM!
